File: matlab/src/solvers/MAM/solver_mna_open.m
function [Q,U,R,T,C,X,lG,totiter] = solver_mna_open(sn, options)

config = options.config;
config.space_max = 32;
if ~isfield(config, 'dep_scv')
    config.dep_scv = 'qna'; % 'qna' for Whitt formula, 'etaqa' for QBD joint moments
end

K = sn.nclasses;
rt = sn.rt;
S = 1./sn.rates;
scv = sn.scv; scv(isnan(scv))=0;

PH = sn.proc;
I = sn.nnodes;
M = sn.nstations;
C = sn.nchains;
V = cellsum(sn.visits);
Q = zeros(M,K);
%QN_1 = Q+Inf;

U = zeros(M,K);
R = zeros(M,K);
T = zeros(M,K);
X = zeros(1,K);

lambda = zeros(1,C);

it = 0;
pie = {};
D0 = {};

% get service process 
for ist=1:M
    switch sn.sched(ist)
        case {SchedStrategy.FCFS, SchedStrategy.INF,SchedStrategy.PS}
            for k=1:K
                pie{ist}{k} = map_pie(PH{ist}{k});
                D0{ist,k} = PH{ist}{k}{1};
                if any(isnan(D0{ist,k}))
                    D0{ist,k} = -GlobalConstants.Immediate;
                    pie{ist}{k} = 1;
                    PH{ist}{k} = map_exponential(GlobalConstants.Immediate);
                end                
            end
    end
end 

a1 = zeros(M,K);
a2 = zeros(M,K);
mubar = [];
c2 = [];
d2 = zeros(M,1);
f2 = zeros(M*K,M*K); 
for ist=1:M
    for jst=1:M
        if sn.nodetype(sn.stationToNode(jst)) ~= NodeType.Source
            for r=1:K
                for s=1:K
                    if rt((ist-1)*K+r, (jst-1)*K+s)>0
                            f2((ist-1)*K+r, (jst-1)*K+s) = 1; % C^2ij,r
                    end
                 end
             end
         end
     end
end   
lambdas_inchain = cell(1,C);
scvs_inchain = cell(1,C);
d2c = [];
for c=1:C
    inchain = sn.inchain{c};
    sourceIdx = sn.refstat(inchain(1));
    lambdas_inchain{c} = sn.rates(sourceIdx,inchain);
    scvs_inchain{c} = scv(sourceIdx,inchain);
    lambda(c) = sum(lambdas_inchain{c}(isfinite(lambdas_inchain{c})));
    d2c(c) = da_traffic_superpos(lambdas_inchain{c},scvs_inchain{c});
    T(sourceIdx,inchain') = lambdas_inchain{c};
   
end
d2(sourceIdx)=d2c(sourceIdx,:)*lambda'/sum(lambda);

%% main iteration
% flow fixed point on the per-station arrival rates a1 and SCVs a2, driven
% by the generic DA successive-substitution driver
fpopts = options;
fpopts.iter_max = options.iter_max + 1; % legacy while-loop executed one extra sweep at the cap
fpopts.config.da_nanstop = true; % legacy while-loop exited on NaN convergence measure
[~, it] = da_fpi(@mna_sweep, [a1(:); a2(:)], fpopts);

    function [xnew, xref] = mna_sweep(~, itnum)
    xref = [a1(:); a2(:)];
    % update throughputs at all stations
    if itnum==1
        for c=1:C
            inchain = sn.inchain{c};
            for m=1:M
                T(m,inchain) = V(m,inchain) .* lambda(c);
            end
        end
    end

    % superposition
    for ist=1:M
        a1(ist,:) = 0;
        a2(ist,:) = 0;
        lambda_i = sum(T(ist,:));        
        for jst=1:M
            for r=1:K
                for s=1:K
                    a1(ist,r) = a1(ist,r) + T(jst,s)*rt((jst-1)*K+s, (ist-1)*K+r);
                    a2(ist,r) = a2(ist,r) + (1/lambda_i) * f2((jst-1)*K+s, (ist-1)*K+r)*T(jst,s)*rt((jst-1)*K+s, (ist-1)*K+r);
                end
            end
        end
    end

    % update flow trhough queueing station
    for ind=1:I
        if sn.isstation(ind)
            ist = sn.nodeToStation(ind);
            
                    switch sn.sched(ist)
                        case SchedStrategy.INF
                                for r=1:K
                                    for s=1:K
                                        d2(ist,s) = a2(ist,s);
                                    end
                            end
                            for c=1:C
                                inchain = sn.inchain{c};
                                for k=inchain
                                    T(ist,k) = a1(ist,k);
                                    U(ist,k) = S(ist,k)*T(ist,k);
                                    Q(ist,k) = T(ist,k).*S(ist,k)*V(ist,k);
                                    R(ist,k) = Q(ist,k)/T(ist,k);
                                end
                            end
                        case SchedStrategy.PS
                            for c=1:C
                                inchain = sn.inchain{c};
                                for k=inchain
                                    TN(ist,k) = lambda(c)*V(ist,k);
                                    UN(ist,k) = S(ist,k)*TN(ist,k);
                                end
                                %Nc = sum(sn.njobs(inchain)); % closed population
                                Uden = min([1-GlobalConstants.FineTol,sum(UN(ist,:))]);
                                for k=inchain
                                    %QN(ist,k) = (UN(ist,k)-UN(ist,k)^(Nc+1))/(1-Uden); % geometric bound type approximation
                                    QN(ist,k) = UN(ist,k)/(1-Uden);
                                    RN(ist,k) = QN(ist,k)/TN(ist,k);
                                end
                            end                    
                        case {SchedStrategy.FCFS}
                            mu_ist = sn.rates(ist,1:K);
                            mu_ist(isnan(mu_ist))=0;                            
                            rho_ist_class = a1(ist,1:K)./(GlobalConstants.FineTol+sn.rates(ist,1:K));
                            rho_ist_class(isnan(rho_ist_class))=0;
                            lambda_ist = sum(a1(ist,:));
                            mi = sn.nservers(ist);
                            rho_ist = sum(rho_ist_class) / mi;
                            if rho_ist < 1-options.tol
                                if strcmp(config.dep_scv, 'etaqa') && mi == 1
                                    % ETAQA-based departure SCV via QBD joint moments
                                    try
                                        % fit aggregate arrival MAP from parametric info
                                        arri_agg = APH.fitMeanAndSCV(1/lambda_ist, sum(a2(ist,:))).getProcess;
                                        serv_agg = APH.fitMeanAndSCV(1/sum(mu_ist(mu_ist>0).*a1(ist,mu_ist>0))/lambda_ist, ...
                                            sum(a1(ist,:).*scv(ist,:))/lambda_ist).getProcess;
                                        JM = qbd_depproc_jointmom(arri_agg, serv_agg, [1,0; 2,0]);
                                        E1 = JM(1); E2 = JM(2);
                                        d2(ist) = (E2 - E1^2) / E1^2;
                                    catch
                                        % fall back to QNA formula on failure
                                        for k=1:K
                                            mubar(ist) = lambda_ist ./ rho_ist;
                                            c2(ist) = -1;
                                            for r=1:K
                                                if mu_ist(r)>0
                                                    c2(ist) = c2(ist) + a1(ist,r)/lambda_ist * (mubar(ist)/mi/mu_ist(r))^2 * (scv(ist,r)+1 );
                                                end
                                            end
                                        end
                                        d2(ist) = 1 + rho_ist^2*(c2(ist)-1)/sqrt(mi) + (1 - rho_ist^2) *(sum(a2(ist,:))-1);
                                    end
                                else
                                    % QNA Whitt formula
                                    for k=1:K
                                        mubar(ist) = lambda_ist ./ rho_ist;
                                        c2(ist) = -1;
                                        for r=1:K
                                            if mu_ist(r)>0
                                                c2(ist) = c2(ist) + a1(ist,r)/lambda_ist * (mubar(ist)/mi/mu_ist(r))^2 * (scv(ist,r)+1 );
                                            end
                                        end
                                    end
                                    d2(ist) = 1 + rho_ist^2*(c2(ist)-1)/sqrt(mi) + (1 - rho_ist^2) *(sum(a2(ist,:))-1);
                                end
                            else
                                for k=1:K
                                    Q(ist,k) = sn.njobs(k);
                                end
                                d2(ist) = 1;
                            end
                            for k=1:K
                                T(ist,k) = a1(ist,k);
                                U(ist,k) = T(ist,k) * S(ist,k) /sn.nservers(ist);
                               
                            end
                    end
            
        else % not a station
            switch sn.nodetype(ind)
                case NodeType.Fork
                    line_error(mfilename,'Fork nodes not supported yet by QNA solver.');
            end
        end
    end    


    % splitting - update flow scvs
    for ist=1:M
        for jst=1:M
            if sn.nodetype(sn.stationToNode(jst)) ~= NodeType.Source
                for r=1:K
                    for s=1:K
                        if rt((ist-1)*K+r, (jst-1)*K+s)>0
                            f2((ist-1)*K+r, (jst-1)*K+s) = 1 + rt((ist-1)*K+r, (jst-1)*K+s) * (d2(ist)-1); 
                        end
                    end
                end
            end
        end
    end   
    xnew = [a1(:); a2(:)];
    end
                

for ind=1:I
        if sn.isstation(ind)
           ist = sn.nodeToStation(ind);
           switch sn.sched(ist)
            case {SchedStrategy.FCFS}
                mu_ist = sn.rates(ist,1:K);
                mu_ist(isnan(mu_ist))=0;                            
                rho_ist_class = a1(ist,1:K)./(GlobalConstants.FineTol+sn.rates(ist,1:K));
                rho_ist_class(isnan(rho_ist_class))=0;
                lambda_ist = sum(a1(ist,:));
                mi = sn.nservers(ist);
                rho_ist = sum(rho_ist_class) / mi;
                if rho_ist < 1-options.tol
                    for k=1:K
                        if a1(ist,k)==0
                            arri_class = map_exponential(Inf);
                         else
                            arri_class = APH.fitMeanAndSCV(1/a1(ist,k),a2(ist,k)).getProcess; % MMAP repres of arrival process for class k at node ist
                            %arri_class = Erlang.fit(1/a1(ist,k),a2(ist,k)).getProcess;
                            %arri_class = map_exponential(1/a1(ist,k));
                            arri_class = {arri_class{1},arri_class{2},arri_class{2}};
                         end
                         if k==1
                            arri_node = arri_class;
                         else
                            arri_node = mmap_super(arri_node,arri_class,  'default');

                            %arri_node = mmap_super_safe({arri_node,arri_class}, config.space_max, 'default'); % combine arrival process from different class
                         end
                    end
                isFiniteCap = isfinite(sn.cap(ist));
                if isFiniteCap
                    capK = sn.cap(ist);
                    [isMmck, muMmck] = mam_detect_mmck(sn, ist, K, arri_node);
                    if isMmck
                        aggrLambda_ist = sum(a1(ist,1:K), 'omitnan');
                        exactRes = qsys_mmck(aggrLambda_ist, muMmck, sn.nservers(ist), capK);
                        meanQ_fc = exactRes.meanQueueLength;
                        lossProb_fc = exactRes.lossProbability;
                    else
                        [meanQ_fc, lossProb_fc, ~] = mam_truncate_renorm( ...
                            {arri_node{[1,3:end]}}, {pie{ist}{:}}, {D0{ist,:}}, capK);
                    end
                    lambdaInflow = a1(ist,1:K);
                    lambdaInflow(isnan(lambdaInflow)) = 0;
                    T_eff = lambdaInflow * (1 - lossProb_fc);
                    sumT = sum(T_eff);
                    if sumT > 0
                        Savg_eff = sum(T_eff .* S(ist,1:K), 'omitnan') / sumT;
                        Wq = max(0, meanQ_fc / sumT - Savg_eff);
                    else
                        Wq = 0;
                    end
                    for k=1:K
                        T(ist,k) = T_eff(k);
                        U(ist,k) = T(ist,k) * S(ist,k) / sn.nservers(ist);
                        if T(ist,k) > 0
                            R(ist,k) = Wq + S(ist,k);
                            Q(ist,k) = T(ist,k) * R(ist,k);
                        else
                            R(ist,k) = 0;
                            Q(ist,k) = 0;
                        end
                    end
                else
                    Qret = cell(1,K);
                    [Qret{1:K}] = MMAPPH1FCFS({arri_node{[1,3:end]}}, {pie{ist}{:}}, {D0{ist,:}}, 'ncMoms', 1);
                    Q(ist,:) = cell2mat(Qret);
                    for k=1:K
                        R(ist,k) = Q(ist,k) ./ T(ist,k);
                    end
                end
                else
                    for k=1:K
                        Q(ist,k) = sn.njobs(k);
                        R(ist,k) = Q(ist,k) ./ T(ist,k);
                    end
                end
            end
            
        end
end

C = sum(R,1);
Q = abs(Q);
Q(isnan(Q))=0;
U(isnan(U))=0;
R(isnan(R))=0;
C(isnan(C))=0;
X(isnan(X))=0;
lG = 0;
totiter = it;
end

File: jar/src/main/java/jline/solvers/mam/handlers/Solver_mna_open.java
package jline.solvers.mam.handlers;

import java.util.HashMap;
import java.util.Map;

import jline.GlobalConstants;
import jline.api.mam.Map_exponential;
import jline.api.mam.Map_pie;
import jline.api.mam.Mmap_shorten;
import jline.api.mam.Mmap_super;
import jline.lang.NetworkStruct;
import jline.lang.constant.NodeType;
import jline.lang.constant.SchedStrategy;
import jline.lang.processes.APH;
import jline.lib.butools.MMAPPH1FCFS;
import jline.solvers.SolverOptions;
import jline.solvers.SolverResult;
import jline.solvers.mam.MAMResult;
import jline.util.Utils;
import jline.api.da.Da_fpi;
import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

public final class Solver_mna_open {
    private Solver_mna_open() {}

    public static SolverResult solver_mna_open(NetworkStruct sn, SolverOptions options) {
        SolverOptions.Config config = options.config;
        config.space_max = 32;
        Object depScvObj = config.get("dep_scv");
        String depScv = (depScvObj instanceof String) ? (String) depScvObj : "qna";

        int K = sn.nclasses;
        Matrix rt = sn.rt.copy();
        Matrix S = sn.rates.elementPow(-1.0);
        Matrix scv = sn.scv.copy();
        scv.removeNaN();
        Map<jline.lang.nodes.Station, Map<jline.lang.JobClass, MatrixCell>> PH = sn.proc;
        int I = sn.nnodes;
        int M = sn.nstations;
        int C = sn.nchains;
        Matrix V = Matrix.cellsum(sn.visits);
        Matrix Q = new Matrix(M, K, M * K);
        Map<Integer, MatrixCell> pie = new HashMap<Integer, MatrixCell>();
        Map<Integer, MatrixCell> DO = new HashMap<Integer, MatrixCell>();

        Matrix U = new Matrix(M, K, M * K);
        Matrix R = new Matrix(M, K, M * K);
        Matrix T = new Matrix(M, K, M * K);
        Matrix X = new Matrix(1, K, K);
        for (int ist = 0; ist < M; ist++) {
            SchedStrategy schedI = sn.sched.get(sn.stations.get(ist));
            if (schedI == SchedStrategy.FCFS || schedI == SchedStrategy.INF || schedI == SchedStrategy.PS) {
                pie.put(ist, new MatrixCell());
                DO.put(ist, new MatrixCell());
                for (int k = 0; k < K; k++) {
                    MatrixCell phProc = PH.get(sn.stations.get(ist)).get(sn.jobclasses.get(k));
                    pie.get(ist).set(k, Map_pie.map_pie(phProc.get(0), phProc.get(1)));
                    Matrix d0Val = phProc.get(0);
                    if (d0Val.hasNaN()) {
                        DO.get(ist).set(k, Matrix.singleton(-GlobalConstants.Immediate));
                        pie.get(ist).set(k, Matrix.singleton(1.0));
                        PH.get(sn.stations.get(ist)).put(sn.jobclasses.get(k), Map_exponential.map_exponential(GlobalConstants.Immediate));
                    } else {
                        DO.get(ist).set(k, d0Val);
                    }
                }
            }
        }

        Matrix lambda = new Matrix(1, C, C);
        int it = 0;
        Matrix a1 = new Matrix(M, K, M * K);
        Matrix a2 = new Matrix(M, K, M * K);

        Matrix d2 = new Matrix(M, 1, M);
        Matrix f2 = new Matrix(M * K, M * K, (int) Math.pow(M * K, 2));
        for (int i = 0; i < M; i++) {
            for (int j = 0; j < M; j++) {
                if (sn.nodetype.get((int) sn.stationToNode.get(j)) != NodeType.Source) {
                    for (int r = 0; r < K; r++) {
                        for (int s = 0; s < K; s++) {
                            if (rt.get(i * K + r, j * K + s) > 0) f2.set(i * K + r, j * K + s, 1);
                        }
                    }
                }
            }
        }
        MatrixCell lambdas_inchain = new MatrixCell();
        MatrixCell scvs_inchain = new MatrixCell();
        Matrix d2c = new Matrix(1, C, C);
        int last_source_idx = 0;
        for (int c = 0; c < C; c++) {
            Matrix inchain = sn.inchain.get(c);
            int sourceIdx = (int) sn.refstat.get((int) inchain.get(0));
            last_source_idx = sourceIdx;
            lambdas_inchain.set(c, new Matrix(1, inchain.length(), inchain.length()));
            for (int i = 0; i < inchain.length(); i++) {
                lambdas_inchain.get(c).set(0, i, sn.rates.get(sourceIdx, (int) inchain.get(i)));
            }
            scvs_inchain.set(c, new Matrix(1, inchain.length(), inchain.length()));
            for (int i = 0; i < inchain.length(); i++) {
                scvs_inchain.get(c).set(0, i, scv.get(sourceIdx, (int) inchain.get(i)));
            }
            Matrix lic = lambdas_inchain.get(c).copy();
            lic.removeInfinite();
            lambda.set(c, lic.elementSum());
            d2c.set(c, Qna_superpos.qna_superpos(lambdas_inchain.get(c), scvs_inchain.get(c)));
            for (int i = 0; i < inchain.length(); i++) {
                T.set(sourceIdx, (int) inchain.get(i), lambdas_inchain.get(c).get(i));
            }
        }
        d2.set(last_source_idx,
                Matrix.extractRows(d2c, last_source_idx, last_source_idx + 1, null).mult(lambda.transpose()).get(0)
                        / lambda.elementSum());
        // flow fixed point on the arrival rates a1 and SCVs a2, driven by the
        // generic DA successive-substitution driver
        Da_fpi.Options<Matrix[]> fpopts = new Da_fpi.Options<Matrix[]>(
                options.iter_max + 1, options.iter_tol, new Da_fpi.Norm<Matrix[]>() {
            @Override
            public double eval(Matrix[] xnew, Matrix[] xref) {
                Matrix d1m = xnew[0].add(-1.0, xref[0]);
                d1m.absEq();
                Matrix d2m = xnew[1].add(-1.0, xref[1]);
                d2m.absEq();
                double e1 = d1m.elementMax();
                double e2 = d2m.elementMax();
                if (e1 > options.iter_tol || e2 > options.iter_tol) {
                    return Double.MAX_VALUE; // keep iterating (legacy OR test)
                } else if (Double.isNaN(e1) || Double.isNaN(e2)) {
                    return Double.NaN;
                }
                return Math.max(e1, e2);
            }
        });
        fpopts.nanstop = true; // legacy while-loop exited on NaN convergence measure
        Da_fpi.Result<Matrix[]> fpres = Da_fpi.run(new Da_fpi.Sweep<Matrix[]>() {
            @Override
            public Da_fpi.SweepResult<Matrix[]> sweep(Matrix[] x, int it) {
            Matrix[] xref = new Matrix[]{a1.copy(), a2.copy()};

            if (it == 1) {
                for (int c = 0; c < C; c++) {
                    Matrix inchain = sn.inchain.get(c);
                    for (int m = 0; m < M; m++) {
                        for (int i = 0; i < inchain.length(); i++) {
                            T.set(m, (int) inchain.get(i), V.get(m, (int) inchain.get(i)) * lambda.get(c));
                        }
                    }
                }
            }
            for (int i = 0; i < M; i++) {
                for (int j = 0; j < K; j++) { a1.set(i, j, 0); a2.set(i, j, 0); }
                double lambda_i = T.sumRows(i);
                for (int j = 0; j < M; j++) {
                    for (int r = 0; r < K; r++) {
                        for (int s = 0; s < K; s++) {
                            a1.set(i, r, a1.get(i, r) + T.get(j, s) * rt.get(j * K + s, i * K + r));
                            a2.set(i, r, a2.get(i, r) + (1 / lambda_i) * f2.get(j * K + s, i * K + r) * T.get(j, s) * rt.get(j * K + s, i * K + r));
                        }
                    }
                }
            }

            for (int ind = 0; ind < I; ind++) {
                if (sn.isstation.get(ind) == 1.0) {
                    int ist = (int) sn.nodeToStation.get(ind);
                    if (sn.nodetype.get(ind) != NodeType.Join) {
                        SchedStrategy schedI = sn.sched.get(sn.stations.get(ist));
                        if (schedI == SchedStrategy.INF) {
                            for (int s = 0; s < K; s++) d2.set(ist, s, a2.get(ist, s));
                            for (int c = 0; c < C; c++) {
                                Matrix inchain = sn.inchain.get(c);
                                for (int k1 = 0; k1 < inchain.length(); k1++) {
                                    int k = (int) inchain.get(k1);
                                    T.set(ist, k, a1.get(ist, k));
                                    U.set(ist, k, S.get(ist, k) * T.get(ist, k));
                                    Q.set(ist, k, T.get(ist, k) * S.get(ist, k) * V.get(ist, k));
                                    R.set(ist, k, Q.get(ist, k) / T.get(ist, k));
                                }
                            }
                        } else if (schedI == SchedStrategy.PS) {
                            for (int c = 0; c < C; c++) {
                                Matrix inchain = sn.inchain.get(c);
                                for (int k1 = 0; k1 < inchain.length(); k1++) {
                                    int k = (int) inchain.get(k1);
                                    T.set(ist, k, lambda.get(c) * V.get(ist, k));
                                    U.set(ist, k, S.get(ist, k) * T.get(ist, k));
                                }
                                double Uden = Math.min(1 - GlobalConstants.FineTol, U.sumRows(ist));
                                for (int k1 = 0; k1 < inchain.length(); k1++) {
                                    int k = (int) inchain.get(k1);
                                    Q.set(ist, k, U.get(ist, k) / (1 - Uden));
                                    R.set(ist, k, Q.get(ist, k) / T.get(ist, k));
                                }
                            }
                        } else if (schedI == SchedStrategy.FCFS) {
                            Matrix mu_ist = new Matrix(1, K, K);
                            for (int i = 0; i < K; i++) mu_ist.set(i, sn.rates.get(ist, i));
                            mu_ist.removeNaN();
                            Matrix rho_ist_class = new Matrix(1, K, K);
                            for (int i = 0; i < K; i++) {
                                rho_ist_class.set(i, a1.get(ist, i) / (GlobalConstants.FineTol + sn.rates.get(ist, i)));
                            }
                            rho_ist_class.removeNaN();
                            double lambda_ist = a1.sumRows(ist);
                            int mi = (int) sn.nservers.get(ist);
                            double rho_ist = rho_ist_class.elementSum() / mi;
                            double c2;
                            if (rho_ist < 1 - options.tol) {
                                double mubar = lambda_ist / rho_ist;
                                c2 = -1.0;
                                for (int r = 0; r < K; r++) {
                                    if (mu_ist.get(r) > 0) {
                                        c2 = c2 + a1.get(ist, r) / lambda_ist * Math.pow(mubar / mi / mu_ist.get(r), 2.0) * (scv.get(ist, r) + 1);
                                    }
                                }
                                d2.set(ist, 1 + Math.pow(rho_ist, 2.0) * (c2 - 1) / Math.sqrt((double) mi)
                                        + (1 - Math.pow(rho_ist, 2.0)) * (a2.sumRows(ist) - 1));
                            } else {
                                for (int k = 0; k < K; k++) Q.set(ist, k, sn.njobs.get(k));
                                d2.set(ist, 1.0);
                            }
                            for (int k = 0; k < K; k++) {
                                T.set(ist, k, a1.get(ist, k));
                                U.set(ist, k, T.get(ist, k) * S.get(ist, k) / sn.nservers.get(ist));
                            }
                        }
                    }
                } else {
                    if (sn.nodetype.get(ind) == NodeType.Fork) {
                        throw new IllegalArgumentException("Fork nodes not supported yet by MNA solver");
                    }
                }
            }

            for (int i = 0; i < M; i++) {
                for (int j = 0; j < M; j++) {
                    if (sn.nodetype.get((int) sn.stationToNode.get(j)) != NodeType.Source) {
                        for (int r = 0; r < K; r++) {
                            for (int s = 0; s < K; s++) {
                                if (rt.get(i * K + r, j * K + s) > 0) {
                                    f2.set(i * K + r, j * K + s, 1 + rt.get(i * K + r, j * K + s) * (d2.get(i) - 1));
                                }
                            }
                        }
                    }
                }
            }
            return new Da_fpi.SweepResult<Matrix[]>(new Matrix[]{a1.copy(), a2.copy()}, xref);
            }
        }, new Matrix[]{a1.copy(), a2.copy()}, fpopts);
        it = fpres.it;

        for (int ind = 0; ind < I; ind++) {
            if (sn.isstation.get(ind) == 1.0) {
                int ist = (int) sn.nodeToStation.get(ind);
                if (sn.sched.get(sn.stations.get(ist)) == SchedStrategy.FCFS) {
                    Matrix mu_ist = new Matrix(1, K, K);
                    for (int i = 0; i < K; i++) mu_ist.set(i, sn.rates.get(ist, i));
                    mu_ist.removeNaN();
                    Matrix rho_ist_class = new Matrix(1, K, K);
                    for (int i = 0; i < K; i++) {
                        rho_ist_class.set(i, a1.get(ist, i) / (GlobalConstants.FineTol + sn.rates.get(ist, i)));
                    }
                    rho_ist_class.removeNaN();
                    int mi = (int) sn.nservers.get(ist);
                    double rho_ist = rho_ist_class.elementSum() / mi;
                    if (rho_ist < 1 - options.tol) {
                        MatrixCell arri_class_anx;
                        MatrixCell arri_class = new MatrixCell();
                        MatrixCell arri_node = new MatrixCell();
                        for (int k = 0; k < K; k++) {
                            if (a1.get(ist, k) == 0.0) {
                                arri_class_anx = Map_exponential.map_exponential(Double.POSITIVE_INFINITY);
                            } else {
                                arri_class_anx = APH.fitMeanAndSCV(1 / a1.get(ist, k), a2.get(ist, k)).getProcess();
                            }
                            arri_class.set(0, arri_class_anx.get(0));
                            arri_class.set(1, arri_class_anx.get(1));
                            arri_class.set(2, arri_class_anx.get(1));
                            if (k == 0) {
                                arri_node.set(0, arri_class.get(0));
                                arri_node.set(1, arri_class.get(1));
                                arri_node.set(2, arri_class.get(2));
                            } else {
                                arri_node = Mmap_super.mmap_super(arri_node, arri_class);
                            }
                        }
                        Map<Integer, Matrix> qretMap = new HashMap<Integer, Matrix>();
                        Object resultObj = MMAPPH1FCFS.solve(Mmap_shorten.mmap_shorten(arri_node),
                                pie.get(ist).toMap(), DO.get(ist).toMap(),
                                1, null, null, null, false, false, null, null).get("ncMoms");
                        if (resultObj instanceof Map) {
                            Map<?, ?> raw = (Map<?, ?>) resultObj;
                            for (Map.Entry<?, ?> e : raw.entrySet()) {
                                if (e.getKey() instanceof Integer && e.getValue() instanceof Matrix) {
                                    qretMap.put((Integer) e.getKey(), (Matrix) e.getValue());
                                }
                            }
                        }
                        for (int i = 0; i < qretMap.size(); i++) {
                            Q.set(ist, i, qretMap.get(i).get(0));
                        }
                    } else {
                        for (int k = 0; k < K; k++) Q.set(ist, k, sn.njobs.get(k));
                    }
                    for (int k = 0; k < K; k++) R.set(ist, k, Q.get(ist, k) / T.get(ist, k));
                }
            }
        }

        MAMResult result = new MAMResult();
        Matrix CN = R.sumCols();
        Q.absEq();
        Q.removeNaN();
        U.removeNaN();
        R.removeNaN();
        CN.removeNaN();
        X.removeNaN();
        result.XN = X;
        result.QN = Q;
        result.CN = CN;
        result.UN = U;
        result.RN = R;
        result.TN = T;
        result.iter = it;
        result.lG = 0.0;
        return result;
    }
}
